### PR TITLE
perf: G1/2 membership using Eval

### PR DESCRIPTION
**Pull request by yelhousni**
# Description

Forgot to push this in #1349. 
G1 and G2 subgroup membership using Eval for curve and twist arithmetics. 

## Type of change

- [x] New feature (non-breaking change which adds functionality)

# How has this been tested?

`TestGroupMembershipSolve` passes.

# How has this been benchmarked?

This saves **101,565 scs** for `GroupMembershipCircuit`.

# Checklist:

- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I did not modify files generated from templates
- [x] `golangci-lint` does not output errors locally
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Review by ivokub — APPROVED** [review]
Looks good in general, I see how you have inlined the Ext2 ops. But have a look at the comment for G1.doubleAndAdd, I don't follow the change of sign.

File: std/algebra/emulated/sw_bls12381/g1.go [hunk 1]
```diff
@@ -108,33 +102,27 @@ func (g1 G1) add(p, q *G1Affine) *G1Affine {
...
 	// omit y1 computation
 	// compute λ1 = -λ1-1*p.y/(x1-p.x)
-	ypyp := g1.curveF.Add(&p.Y, &p.Y)
+	ypyp := g1.curveF.MulConst(&p.Y, big.NewInt(2))
 	x2xp := g1.curveF.Sub(x2, &p.X)
 	λ2 := g1.curveF.Div(ypyp, x2xp)
 	λ2 = g1.curveF.Add(λ1, λ2)
-	λ2 = g1.curveF.Neg(λ2)
 
 	// compute x3 =λ2²-p.x-x3
```
[the remark this answers is not in the document]
Comment (yelhousni) [reply std/algebra/emulated/sw_bls12381/g1.go:121]: should be `(-λ2)²-p.x-x3`

File: std/algebra/emulated/sw_bls12381/g1.go [hunk 2]
```diff
@@ -108,33 +102,27 @@ func (g1 G1) add(p, q *G1Affine) *G1Affine {
...
-	λ2λ2 := g1.curveF.Mul(λ2, λ2)
-	x3 := g1.curveF.Sub(λ2λ2, &p.X)
-	x3 = g1.curveF.Sub(x3, x2)
+	x3 := g1.curveF.Eval([][]*baseEl{{λ2, λ2}, {mone, &p.X}, {mone, x2}}, []int{1, 1, 1})
 
 	// compute y3 = λ2*(p.x - x3)-p.y
```
[the remark this answers is not in the document]
Comment (yelhousni) [reply std/algebra/emulated/sw_bls12381/g1.go:124]: now we return `- slope` in `λ2`. Will correct comments.

[the remark this answers is not in the document]
Comment (ivokub) [reply std/algebra/emulated/sw_bls12381/g1.go:124]: Ah, I see! Missed the negation of lambda2

**Pull request merged**